# Data and scripts used in: Mechanisms of carbon dioxide detection in the earthworm _Dendrobaena veneta_.

### E Jordan Smith, Jennifer L Ryan, Sofia A Lopresti, Dastan BS Haghnazari, Karleigh AS Anderson, Sarah J Lipson, Erik C Johnson, Wayne L Silver, Cecil J Saunders

Preprint available on Biorxiv: [https://www.biorxiv.org/content/10.1101/2023.04.12.536649v1](https://www.biorxiv.org/content/10.1101/2023.04.12.536649v1)

###### Folder Contents:

**[DEseq_ggtrees_scripts](https://github.com/JakeSaunders/Smith_etal_2023_Earthworm_CO2/tree/main/DEseq_ggtrees_scripts)**

R scripts and requisite files used to conduct analysis of transcript expression and produce trees featured in Smith *et al*, 2023.

**[Transcriptome_Assembly_Trinity_Trinotate](https://github.com/JakeSaunders/Smith_etal_2023_Earthworm_CO2/tree/main/Transcriptome_Assembly_Trinity_Trinotate)** 

Slurm scripts that assemble and annotate _de novo_ transcriptome from raw Illumina reads using [Trinity](https://github.com/trinityrnaseq/trinityrnaseq/wiki) and [Trinotate](https://github.com/Trinotate/Trinotate/wiki). Raw RNA-sequencing data can be found at NCBI's Sequence Read Archive (SRA) under the bioproject ID PRJNA1026068. 

[exudate_assay_data_stats](https://github.com/JakeSaunders/Smith_etal_2023_Earthworm_CO2/tree/main/exudate_assay_data_stats)

Raw exudate assay data in csv format and an R script to generate graphs and statistical analysis of data featured in Smith *et al*, 2023.
